### DISCUSSION TITLE
Kornia enhancement on very large image tensors

### DISCUSSION THREAD

**Discussion by mpelchat04** [kornia/kornia#1927] [Ideas]
We're currently working with very large images that do not fit into the GPU memory, for which we'd like to use the CLAHE enhancement.   
## Problem statement:  
Creating smaller patches of the image and applying the enhancement per patch creates unwanted artifacts. For example, a patch of low varying pixel values will be overly-enhanced, compared to other patches from the complete image.
We're currently using the Scikit-learn CLAHE implementation (which gives fine results but does not make use of the GPU) but we'd like to move toward the Kornia one, to improve the speed of this.  

## Questions:
Would it be feasible to apply the Kornia implementation of the CLAHE, per patch, using the pixel distribution information from the whole image?  For example, we could pre-calculate and provide an histogram for the whole image).   

Would it require important modifications to the Kornia code base?

**Comment by lferraz** [comment]
The current code of CLAHE is based on the OpenCV implementation which contains some tricks to speed up the process. Probably for this reason the result is not the same as SciKit, anyway please compare the results with OpenCV.

As far as I remember is not difficult to modify the code to get the same results as the original CLAHE algorithm but I do not remember exactly. Not sure how difficult can be injecting a histogram, I suppose it can be posible with a bypass of the function that computes the histograms.

Anyway, for sure, the internal code needs to be changed so, I think it is easier to add a flag in the method to select nice results vs fast.

**Reply by lferraz** [reply]
Related with the original question: if you compute a whole histogram then you are applying a differnet algorithm. As far as I remember the advantage of CLAHE is that takes into account local histograms. If you want to use a global histogram then check the [`equalize`](https://kornia.readthedocs.io/en/v0.5.0/_modules/kornia/enhance/adjust.html#equalize) function. The code is much much simpler.

**Reply by edgarriba** [reply]
would be nice to see a qualitative comparison between patched clahe and patched global equalize
